File: TrinaryCipher.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#*********************************************************
# 作者：伍耀晖               Author: YaoHui.Wu           *
# 开源日期：2022年5月27日    Open Source Date: 2022-5-27 *
# 国家：中国                 Country: China              *
#*********************************************************

import sys

def Usage():
    print("Usage\n\tEncryption: python TrinaryCipher -e/-E Plaintext.file Ciphertext.file Password\n\tDecrytion: python TrinaryCipher -d/-D Ciphertext.file Plaintext.file Password")

def Ternary(iNumeric, lTrinary):
    for i in range(5, -1, -1):
        iNumeric, iRemainder = divmod(iNumeric, 3)

        lTrinary[i] = iRemainder

if __name__ == "__main__":
    if len(sys.argv) != 5: Usage()

    elif sys.argv[1] == "-e" or sys.argv[1] == "-E":
        with open(sys.argv[2], "br") as fdPlaintext:
            bData = fdPlaintext.read()

            iFileSize = fdPlaintext.tell()

        if iFileSize == 0:
            print("There is no data in file [{}], 0 byte.".format(sys.argv[2]))

            exit(-1)

        strPassword, lData = sys.argv[4], []

        iPasswordLength, k = len(strPassword), 0

        lPlaintextOrCiphertext, lPassword = [0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0]

        for i in range(iFileSize):
            Ternary(bData[i], lPlaintextOrCiphertext)

            Ternary(ord(strPassword[k]), lPassword)

            for j in range(6):
                if lPlaintextOrCiphertext[j] == lPassword[j] == 0: lPlaintextOrCiphertext[j] = "0"

                elif lPlaintextOrCiphertext[j] == 0 and lPassword[j] == 1: lPlaintextOrCiphertext[j] = "0"

                elif lPlaintextOrCiphertext[j] == 0 and lPassword[j] == 2: lPlaintextOrCiphertext[j] = "2"

                elif lPlaintextOrCiphertext[j] == 1 and lPassword[j] == 0: lPlaintextOrCiphertext[j] = "1"

                elif lPlaintextOrCiphertext[j] == lPassword[j] == 1: lPlaintextOrCiphertext[j] = "1"

                elif lPlaintextOrCiphertext[j] == 1 and lPassword[j] == 2: lPlaintextOrCiphertext[j] = "1"

                elif lPlaintextOrCiphertext[j] == 2 and lPassword[j] == 0: lPlaintextOrCiphertext[j] = "2"

                elif lPlaintextOrCiphertext[j] == 2 and lPassword[j] == 1: lPlaintextOrCiphertext[j] = "2"

                elif lPlaintextOrCiphertext[j] == lPassword[j] == 2: lPlaintextOrCiphertext[j] = "0"

            lData.append(int("".join(lPlaintextOrCiphertext), 3))

            k = (k + 1) % iPasswordLength

        with open(sys.argv[3], "bw") as fdCiphertext:
            for iData in lData:
                fdCiphertext.write(iData.to_bytes(2, "little"))

    elif sys.argv[1] == "-d" or sys.argv[1] == "-D":
        with open(sys.argv[2], "br") as fdCiphertext:
            bData = fdCiphertext.read()

            iFileSize = fdCiphertext.tell() // 2

        if iFileSize == 0:
            print("There is no data in file [{}], 0 byte.".format(sys.argv[2]))

            exit(-1)

        strPassword, lData = sys.argv[4], []

        iPasswordLength, k = len(strPassword), 0

        lCiphertextOrPlaintext, lPassword = [0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0]

        for i in range(iFileSize):
            Ternary(bData[2 * i] + (bData[2 * i +1] << 8), lCiphertextOrPlaintext)

            Ternary(ord(strPassword[k]), lPassword)

            for j in range(6):
                if lCiphertextOrPlaintext[j] == lPassword[j] == 0: lCiphertextOrPlaintext[j] = "0"

                elif lCiphertextOrPlaintext[j] == 0 and lPassword[j] == 1: lCiphertextOrPlaintext[j] = "0"

                elif lCiphertextOrPlaintext[j] == 0 and lPassword[j] == 2: lCiphertextOrPlaintext[j] = "2"

                elif lCiphertextOrPlaintext[j] == 1 and lPassword[j] == 0: lCiphertextOrPlaintext[j] = "1"

                elif lCiphertextOrPlaintext[j] == lPassword[j] == 1: lCiphertextOrPlaintext[j] = "1"

                elif lCiphertextOrPlaintext[j] == 1 and lPassword[j] == 2: lCiphertextOrPlaintext[j] = "1"

                elif lCiphertextOrPlaintext[j] == 2 and lPassword[j] == 0: lCiphertextOrPlaintext[j] = "2"

                elif lCiphertextOrPlaintext[j] == 2 and lPassword[j] == 1: lCiphertextOrPlaintext[j] = "2"

                elif lCiphertextOrPlaintext[j] == lPassword[j] == 2: lCiphertextOrPlaintext[j] = "0"

            lData.append(int("".join(lCiphertextOrPlaintext), 3))

            k = (k + 1) % iPasswordLength

        with open(sys.argv[3], "bw") as fdPlaintext:
            for iData in lData:
                fdPlaintext.write(iData.to_bytes(1, "little"))

    else: Usage()

File: TrinaryCipher.java
/*********************************************************
* 作者：伍耀晖               Author: YaoHui.Wu           *
* 开源日期：2022年5月27日    Open Source Date: 2022-5-27 *
* 国家：中国                 Country: China              *
*********************************************************/

import java.io.*;

public class TrinaryCipher
{
    private static void Usage()
    {
        System.out.println("Usage\n\tEncryption: java TrinaryCipher -e/-E Plaintext.file Ciphertext.file Password\n\tDecryption: java TrinaryCipher -d/-D Ciphertext.file Plaintext.file Password");
    }

    private static void Ternary(int iNumeric,
                                byte[] baTrinary)
    {
        baTrinary[0] = baTrinary[1] = baTrinary[2] = baTrinary[3] = baTrinary[4] = baTrinary[5] =  0;

        if(iNumeric != 0)
        {
            for(int i = 5; i > -1; --i)
            {
                baTrinary[i] = (byte)(iNumeric % 3);

                iNumeric /= 3;
            }
        }
    }

    public static void main(String[] args)
    {
        if(args.length != 4)
        {
            Usage();
        }
        else if(args[0].equals("-e") || args[0].equals("-E"))
        {
            File fdData = new File(args[1]);

            long lFileSize = fdData.length();

            if(lFileSize == 0)
            {
                System.out.printf("There is no data in file [%s], 0 byte.\n", args[1]);

                System.exit(-1);
            }

            byte[] baPlaintext = new byte[(int)lFileSize], baCiphertext = new byte[(int)(2 * lFileSize)];

            try
            {
                FileInputStream fisData = new FileInputStream(fdData);

                fisData.read(baPlaintext, 0, (int)lFileSize);

                fisData.close();
            }
            catch(Exception e)
            {
                e.printStackTrace();
            }

            int iPasswordLength = args[3].length();

            byte[] baPlaintextOrCiphertext = new byte[6], baPassword = new byte[6];

            for(int i = 0, k = 0; i < lFileSize; ++i, k = ++k % iPasswordLength)
            {
                Ternary(baPlaintext[i] & 255, baPlaintextOrCiphertext);

                Ternary(args[3].getBytes()[k], baPassword);

                for(int j = 0; j < 6; ++j)
                {
                    if(baPlaintextOrCiphertext[j] == 0 && baPassword[j] == 0)
                    {
                        baPlaintextOrCiphertext[j] = 0;
                    }
                    else if(baPlaintextOrCiphertext[j] == 0 && baPassword[j] == 1)
                    {
                        baPlaintextOrCiphertext[j] = 0;
                    }
                    else if(baPlaintextOrCiphertext[j] == 0 && baPassword[j] == 2)
                    {
                        baPlaintextOrCiphertext[j] = 2;
                    }
                    else if(baPlaintextOrCiphertext[j] == 1 && baPassword[j] == 0)
                    {
                        baPlaintextOrCiphertext[j] = 1;
                    }
                    else if(baPlaintextOrCiphertext[j] == 1 && baPassword[j] == 1)
                    {
                        baPlaintextOrCiphertext[j] = 1;
                    }
                    else if(baPlaintextOrCiphertext[j] == 1 && baPassword[j] == 2)
                    {
                        baPlaintextOrCiphertext[j] = 1;
                    }
                    else if(baPlaintextOrCiphertext[j] == 2 && baPassword[j] == 0)
                    {
                        baPlaintextOrCiphertext[j] = 2;
                    }
                    else if(baPlaintextOrCiphertext[j] == 2 && baPassword[j] == 1)
                    {
                        baPlaintextOrCiphertext[j] = 2;
                    }
                    else if(baPlaintextOrCiphertext[j] == 2 && baPassword[j] == 2)
                    {
                        baPlaintextOrCiphertext[j] = 0;
                    }
                }

                char cCiphertext = (char)(243 * baPlaintextOrCiphertext[0] + 81 * baPlaintextOrCiphertext[1] + 27 * baPlaintextOrCiphertext[2] + 9 * baPlaintextOrCiphertext[3] + 3 * baPlaintextOrCiphertext[4] + baPlaintextOrCiphertext[5]);

                baCiphertext[2 * i] = (byte)cCiphertext;

                baCiphertext[2 * i + 1] = (byte)(cCiphertext >> 8);
            }
            
            try
            {
                FileOutputStream fosData = new FileOutputStream(args[2]);

                fosData.write(baCiphertext, 0, (int)(2 * lFileSize));

                fosData.close();
            }
            catch(Exception e)
            {
                e.printStackTrace();
            }
        }
        else if(args[0].equals("-d") || args[0].equals("-D"))
        {
            File fdData = new File(args[1]);

            long lFileSize = fdData.length();

            if(lFileSize == 0)
            {
                System.out.printf("There is no data in file [%s], 0 byte.\n", args[1]);

                System.exit(-1);
            }

            byte[] baCiphertext = new byte[(int)lFileSize];

            try
            {
                FileInputStream fisData = new FileInputStream(fdData);

                fisData.read(baCiphertext, 0, (int)lFileSize);

                fisData.close();
            }
            catch(Exception e)
            {
                e.printStackTrace();
            }

            lFileSize /= 2;

            byte[] baPlaintext = new byte[(int)lFileSize];

            int iPasswordLength = args[3].length();

            byte[] baCiphertextOrPlaintext = new byte[6], baPassword = new byte[6];

            for(int i = 0, k = 0; i < lFileSize; ++i, k = ++k % iPasswordLength)
            {
                Ternary((baCiphertext[2 * i] & 255) + ((baCiphertext[2 * i + 1] & 255) << 8), baCiphertextOrPlaintext);

                Ternary(args[3].getBytes()[k], baPassword);

                for(int j = 0; j < 6; ++j)
                {
                    if(baCiphertextOrPlaintext[j] == 0 && baPassword[j] == 0)
                    {
                        baCiphertextOrPlaintext[j] = 0;
                    }
                    else if(baCiphertextOrPlaintext[j] == 0 && baPassword[j] == 1)
                    {
                        baCiphertextOrPlaintext[j] = 0;
                    }
                    else if(baCiphertextOrPlaintext[j] == 0 && baPassword[j] == 2)
                    {
                        baCiphertextOrPlaintext[j] = 2;
                    }
                    else if(baCiphertextOrPlaintext[j] == 1 && baPassword[j] == 0)
                    {
                        baCiphertextOrPlaintext[j] = 1;
                    }
                    else if(baCiphertextOrPlaintext[j] == 1 && baPassword[j] == 1)
                    {
                        baCiphertextOrPlaintext[j] = 1;
                    }
                    else if(baCiphertextOrPlaintext[j] == 1 && baPassword[j] == 2)
                    {
                        baCiphertextOrPlaintext[j] = 1;
                    }
                    else if(baCiphertextOrPlaintext[j] == 2 && baPassword[j] == 0)
                    {
                        baCiphertextOrPlaintext[j] = 2;
                    }
                    else if(baCiphertextOrPlaintext[j] == 2 && baPassword[j] == 1)
                    {
                        baCiphertextOrPlaintext[j] = 2;
                    }
                    else if(baCiphertextOrPlaintext[j] == 2 && baPassword[j] == 2)
                    {
                        baCiphertextOrPlaintext[j] = 0;
                    }
                }

                baPlaintext[i] = (byte)(243 * baCiphertextOrPlaintext[0] + 81 * baCiphertextOrPlaintext[1] + 27 * baCiphertextOrPlaintext[2] + 9 * baCiphertextOrPlaintext[3] + 3 * baCiphertextOrPlaintext[4] + baCiphertextOrPlaintext[5]);
            }

            try
            {
                FileOutputStream fosData = new FileOutputStream(args[2]);

                fosData.write(baPlaintext, 0, (int)lFileSize);

                fosData.close();
            }
            catch(Exception e)
            {
                e.printStackTrace();
            }
        }
        else
        {
            Usage();
        }
    }
}